src/genomic_features/ensembl/ensembldb.py

```diff
@@ -127,3 +128,53 @@ def genes(
     def chromosomes(self):
         """Get chromosome information."""
         return self.db.table("chromosome").execute()
+
+    def promoters(
+        self,
+        filter: _filters.AbstractFilterExpr = filters.EmptyFilter(),
+        upstream: int = 2000,
+        downstream: int = 200,
+        canonical_transcripts: bool = False,
+    ) -> DataFrame:
+        """Get promoter annotations.
+
+        Parameters
+        ----------
+        filter
+            Filter expression to apply to the genes table.
+        upstream
+            Number of base pairs upstream of the TSS (default: 2000).
+        downstream
+            Number of base pairs downstream of the TSS (default: 200).
+        canonical_transcripts
+            If True, return only canonical transcript for each gene (default: False).
+
+        Returns
+        -------
+        DataFrame
+            A table of promoter annotations.
+        """
+        # TODO: change to get transcript table with gene level columns
+        # something like:
+        # tx_table = self.transcripts(cols = set(cols + ['seq_strand', 'seq_name', 'tx_is_canonical']), filter = filter)
+        tx_table = self.genes(filter)
+
+        # Get promoter region based on strand
+        # strand = 1         |>>>>>>>>>>>>>>|
+        # strand = -1                         |<<<<<<<<<<<<<<|
+        # Tx SS:             *                               *
+        # Promoter       ------                             ------
+        tx_ss = np.where(
+            tx_table["seq_strand"] == 1,
+            tx_table["gene_seq_start"],
+            tx_table["gene_seq_end"],
+        )
+        tx_table["promoter_seq_start"] = np.where(
+            tx_table["seq_strand"] == 1, tx_ss - upstream, tx_ss - downstream
+        )
+        tx_table["promoter_seq_end"] = np.where(
+            tx_table["seq_strand"] == 1, tx_ss + downstream, tx_ss + upstream
+        )
```

Comment:
Nice! Would you be up for PR-ing this into bioframe as well? https://github.com/open2c/bioframe/issues/144

Reply:
Drafted in https://github.com/open2c/bioframe/pull/152